derp/derp_server.go

```diff
@@ -889,6 +981,96 @@ func (c *sclient) sendPacket(srcKey key.Public, contents []byte) (err error) {
 	return err
 }
 
+// AddPacketForwarder registers fwd as a packet forwarder for dst.
+// fwd must be comparable.
+func (s *Server) AddPacketForwarder(dst key.Public, fwd PacketForwarder) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.clientsElsewhere == nil {
+		s.clientsElsewhere = map[key.Public]PacketForwarder{}
+	}
+	if prev, ok := s.clientsElsewhere[dst]; ok {
+		if prev == fwd {
+			// Duplicate registration of same forwarder. Ignore.
+			return
+		}
+		if m, ok := prev.(multiForwarder); ok {
+			if _, ok := m[fwd]; !ok {
+				// Duplicate registration of same forwarder in set; ignore.
+				return
+			}
+			m[fwd] = m.maxVal() + 1
+			return
+		}
+		// Otherwise, the existing value is not a set and not a dup, so make it a set.
+		fwd = multiForwarder{
+			prev: 1,
```

Comment:
prev=1 fwd=2 reads a bit mysterious. I expect it'll make sense once I reach the implementation of multiForwarder, but in this reading order, it's confusing.

Reply:
Also: where does the passed-in fwd go? I expected multiForwarder to grab a reference to it or something, but it just gets dropped on the floor?

Reply:
Oh, multiForwarder is a map, and fwd is a ref to the variable, not a struct key. Hrm. This reads very confusing without knowing beforehand that this is a map. Maybe make it more explicit with:

```
multi = multiForwarder{}
multi[prev] = 1
multi[fwd] = 2
```

Or something like that? Maybe it's just me that finds this confusing, no strong feeling here.